docs/src/lib/indexing.md

```diff
@@ -21,77 +21,146 @@ The rules for a valid type of index into a column are the following:
+* `@view sdf[row, cols]` -> a `DataFrameRow` with parent `parent(sdf)` if `cols` is a colon and `parent(sdf)[!, cols]` otherwise;
+* `@view sdf[rows, col]` -> translates to `view(sdf[!, col], rows)` (a standard view into `sdf[!, col]` vector);
+* `@view sdf[rows, cols]` -> a `SubDataFrame` with parent `parent(sdf)` if `cols` is a colon and `parent(sdf)[!, cols]` otherwise;
+* `@view sdf[:, col]` -> the same as `sdf[!, col]`;
+* `@view sdf[:, cols]` -> the same as `sdf[!, cols]`;
+* `@view sdf[!, col]` -> the same as `sdf[!, col]`;
+* `@view sdf[!, cols]` -> the same as `sdf[!, cols]`.
+
+`getindex` on `DataFrameRow`:
+* `dfr[col]` -> the value contained in column `col` of `dfr`; the same as `dfr.col` is `col` is a valid identifier;
 * `dfr[cols]` -> a `DataFrameRow` with parent `parent(dfr)` if `cols` is a colon and `parent(dfr)[cols]` otherwise;
+
+`view` on `DataFrameRow`:
 * `@view dfr[col]` -> a `0`-dimensional view into `parent(dfr)[DataFrames.row(dfr), col]`;
 * `@view dfr[cols]` -> a `DataFrameRow` with parent `parent(dfr)` if `cols` is a colon and `parent(dfr)[cols]` otherwise;
 
+Note that when views are created with columns selector set to `:` then the view produced dynamically changes its columns'
+names and count if columns of parents are added/removed/renamed;
+
 ## `setindex!`
 
-Under construction
+The following list specifies return types of `setindex!` operations depending on argument types.
+
+In particular a description explicitly mentions if the assignment is *in-place*,
+*replaces old vectors without copying source* or *replaces old vectors with copying source*.
+
+`setindex!` on `DataFrame`:
+* `df[CartesianIndex(row, col)] = v` -> the same as `df[row, col] = v`;
+* `df[row, col] = v` -> set value of `col` in row `row` to `v` in-place;
+* `df[row, cols] = v` -> the same as `dfr = df[row, cols]; dfr[:] = v` in-place;
+* `df[rows, col] = v` -> set rows `rows` of column `col` in-place; `v` can be an abstract vector;
+* `df[rows, cols]` -> set rows `rows` of columns `cols` in-place; `v` can be an `AbstractMatrix` or `v` can be `AbstractDataFrame`
+                      (in this case column names must match);
+* `df[:, col] = v` -> the same as `df[1:nrow(df), col] = v` but replaces old vectors with a copy of source;
+                      the same as `df.col = v` if `col` is a valid identifier;
+                      also if `col` is a `Symbol` that is not present in `df` then a new column in `df` is created;
+* `df[:, cols] = v` -> the same as `df[1:nrow(df), cols] = v` but replaces old vectors with a copy of source;
+* `df[!, col] = v` -> the same as `df[1:nrow(df), col] = v` but replaces old vectors without copying source;
+                      also if `col` is a `Symbol` that is not present in `df` then a new column in `df` is created without copying of source;
+* `df[!, cols] = v` -> the same as `df[1:nrow(df), cols] = v` but replaces old vectors without copying source if `v` is an `AbstractDataFrame`.
```

Comment:
So this means no copy is made for `SubDataFrame`?

Reply:
This rule is for `df`. For `sdf` there is a separate rule below `sdf[!, cos] = v` is an in-place operation.

Reply:
I'm referring to
> if `v` is an `AbstractDataFrame`.

It would be weird when `v` is a `SubDataFrame` to set columns of `df` to be `SubArray`s. OTOH maybe people shouldn't use `!` in that case. Not sure.

Reply:
I agree it is weird. But the same will happen with
```
foreach(col -> df[!, col] = sdf[!, col], axes(df, 2))
```
which is essentially the same operation (but without checking that the column names match).

The reason for this problem is that we allow to store in `DataFrame` any `AbstractVector` and not restrict it to `Vector`.